minetest/irrlicht
# Refactor parts of the material system and set min filter to bilinear for 2D graphics

This is the Irrlicht side of minetest/minetest#13622.

This PR fixes minetest/minetest#13621. Because it exposes all OpenGL texture filtering options to Minetest, it also makes minetest/minetest#13108 trivial to fix.

### Part 1: Refactor parts of the material system

The first 6 commits of this PR are the "Refactor parts of the Irrlicht material system" part.

Before this PR:

- `SMaterialLayer` has two boolean fields, `BilinearFilter` and `TrilinearFilter`, which select both the min and the mag filter
- you set them via `SMaterial::setFlag`, which automatically sets them on all texture layers for you

After this PR:

- `SMaterialLayer` has two enum fields, `MinFilter` and `MagFilter`, which select the min and the mag filter
- ❌ setting them via `SMaterial::setFlag`? impossible because they are no booleans
- ❌ manually setting them on all texture layers? would cause code duplication
- ✅ you set them in a lambda passed to the new `SMaterial::forEachTexture` function

However, setting a per-texture material flag on all texture layers was the only real use of `SMaterial::setFlag`. All other uses could trivially be replaced by something like `material.<name of the flag> = <value>`, i.e. `setFlag` was a simple setter function. Therefore, I replaced all of its uses and removed it.

There also was `ISceneNode::setMaterialFlag`, which set a material flag on all materials of a scene node for you. Because I needed a replacement for that as well, I introduced `ISceneNode::forEachMaterial`.

One thing I like about the new API is that it makes it explicit at what layer you are configuring something right now. Look at this code using the old API:

```c++
// sets the type of all materials of this scene node
m_spritenode->setMaterialType(video::EMT_TRANSPARENT_ALPHA_CHANNEL);
// sets a flag on all materials of this scene node
m_spritenode->setMaterialFlag(video::EMF_LIGHTING, false);
// sets the first texture of all materials of this scene node
m_spritenode->setMaterialTexture(0,
		env->getGameDef()->tsrc()->getTextureForMesh("smoke_puff.png"));
// sets a flag on all textures of all materials of this scene node
m_spritenode->setMaterialFlag(video::EMF_BILINEAR_FILTER, false);
```

Without the comments, would you have guessed what each function call does? This is the equivalent code with the new API:

```c++
video::ITexture *smoke = env->getGameDef()->tsrc()->getTextureForMesh("smoke_puff.png");

m_spritenode->forEachMaterial([smoke] (video::SMaterial &mat) {
	mat.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
	mat.Lighting = false;
	mat.setTexture(0, smoke);
	mat.forEachTexture([] (video::SMaterialLayer &tex) {
		tex.BilinearFilter = false;
	});
});
```

The new API makes it obvious that you should actually be doing this:

```c++
video::ITexture *smoke = env->getGameDef()->tsrc()->getTextureForMesh("smoke_puff.png");

m_spritenode->forEachMaterial([smoke] (video::SMaterial &mat) {
	mat.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
	mat.Lighting = false;
	mat.setTexture(0, smoke);
	mat.TextureLayers[0].BilinearFilter = false;
});
```

(The two "new API" examples don't take the switch from `BilinearFilter` and `TrilinearFilter` to `MinFilter` and `MagFilter` into account.)

### Part 2: Set min filter to bilinear for 2D

The last commit of this PR is the "Set min filter to bilinear for 2D graphics" part.

This commit enables bilinear filtering for downscaling textures in Irrlicht's 2D renderer (which is used by Minetest's GUIs). Essentially, it just sets `GL_TEXTURE_MIN_FILTER` to `GL_LINEAR` for 2D rendering.

## To do

This PR is a Ready for Review.

## How to test

Use Minetest on all platforms and with all Irrlicht video drivers and verify that everything looks the same as before, with one exception: Images rendered with a lower resolution than their original resolution in GUIs look better. The difference is very visible with minetest/minetest#13606, for example.

## Additional context

Minetest cannot upscale GUI images with bilinear filtering because that would blur pixel art, but we can _downscale_ images with bilinear filtering. This makes downscaled images look much better. Compare these screenshots:

\[1\] before this PR, \[2\] after this PR

![screenshot before](https://github.com/minetest/irrlicht/assets/82708541/9ae2bf5b-bef2-4930-a50d-3d231cc4d1a0) ![screenshot after](https://github.com/minetest/irrlicht/assets/82708541/27f69dd3-431f-4bf8-9c98-9c2fa9f7523e)
